api/api_ethereum.go

```diff
@@ -1029,3 +1282,226 @@ func (api *EthereumAPI) rpcMarshalBlock(block *types.Block, inclTx, fullTx bool)
+		return 0, errors.New("both gasPrice and (maxFeePerGas or maxPriorityFeePerGas) specified")
+	} else if args.GasPrice != nil {
+		feeCap = args.GasPrice.ToInt()
+	} else if args.MaxFeePerGas != nil {
+		feeCap = args.MaxFeePerGas.ToInt()
+	} else {
+		feeCap = common.Big0
+	}
+	// recap the highest gas limit with account's available balance.
+	if feeCap.BitLen() != 0 {
+		state, _, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+		if err != nil {
+			return 0, err
+		}
+		balance := state.GetBalance(*args.From) // from can't be nil
+		available := new(big.Int).Set(balance)
+		if args.Value != nil {
+			if args.Value.ToInt().Cmp(available) >= 0 {
+				return 0, errors.New("insufficient funds for transfer")
+			}
+			available.Sub(available, args.Value.ToInt())
+		}
+		allowance := new(big.Int).Div(available, feeCap)
+
+		// If the allowance is larger than maximum uint64, skip checking
+		if allowance.IsUint64() && hi > allowance.Uint64() {
+			transfer := args.Value
+			if transfer == nil {
+				transfer = new(hexutil.Big)
+			}
+			logger.Warn("Gas estimation capped by limited funds", "original", hi, "balance", balance,
+				"sent", transfer.ToInt(), "maxFeePerGas", feeCap, "fundable", allowance)
+			hi = allowance.Uint64()
+		}
+	}
+	// Recap the highest gas allowance with specified gascap.
+	if gasCap != 0 && hi > gasCap {
+		logger.Warn("Caller gas above allowance, capping", "requested", hi, "cap", gasCap)
+		hi = gasCap
+	}
+	cap = hi
+
+	// Create a helper to check if a gas allowance results in an executable transaction.
+	// executable returns
+	// - bool: true when a call with given args and gas is executable.
+	// - []byte: EVM execution result.
+	// - error: error occurred during EVM execution.
+	// - error: consensus error which is not EVM related error (less balance of caller, wrong nonce, etc...).
+	executable := func(gas uint64) (bool, []byte, error, error) {
+		args.Gas = (*hexutil.Uint64)(&gas)
+		ret, _, err := EthDoCall(ctx, b, args, rpc.NewBlockNumberOrHashWithNumber(rpc.LatestBlockNumber), nil, 0, gasCap)
+		if err != nil {
+			if isVMError(err) {
+				// If err is vmError, return vmError with returned data
+				return false, ret, err, nil
+			}
+			// Returns error when it is not VM error (less balance or wrong nonce, etc...).
+			return false, nil, nil, err
+		}
+		return true, ret, nil, nil
+	}
+	// Execute the binary search and hone in on an executable gas limit
+	for lo+1 < hi {
+		mid := (hi + lo) / 2
+		isExecutable, _, _, err := executable(mid)
+		if err != nil {
+			return 0, err
+		}
+
+		if !isExecutable {
+			lo = mid
+		} else {
+			hi = mid
+		}
+	}
+	// Reject the transaction as invalid if it still fails at the highest allowance
+	if hi == cap {
+		isExecutable, ret, vmErr, err := executable(hi)
+		if err != nil {
+			return 0, err
+		}
+		if !isExecutable {
+			if ret != nil {
+				if vmErr == vm.ErrExecutionReverted || strings.Contains(vmErr.Error(), vm.ErrExecutionReverted.Error()) {
```

Comment:
This is just question, can you explain why did you add `strings.Contains(vmErr.Error(), vm.ErrExecutionReverted.Error())` to check `vmErr` is `vm.ErrExecutionReverted`?
Is there any logic to wrap vm error in `EthDoCall`?

Reply:
Please don't mind this comment